go/cmd/server/logging/logging.go

```diff
@@ -0,0 +1,306 @@
+	ticker := time.NewTicker(100 * time.Millisecond)
+	defer ticker.Stop()
+
+	for {
+		s.ProcessMemoryBuffer(ticker)
+	}
+}
+
+// Select that eitheringests new logs that are added to the logging channel, one at a time to add
+// to the in memory buffer or flushes all of them synchronously to the OfflineStorage on a time interval.
+func (s *LoggingService) ProcessMemoryBuffer(t *time.Ticker) {
+	select {
+	case t := <-t.C:
+		s.flushLogsToOfflineStorage(t)
+	case new_log := <-s.logChannel:
+		log.Printf("Adding %s to memory.\n", new_log.FeatureValues)
+		s.memoryBuffer.logs = append(s.memoryBuffer.logs, new_log)
+	}
+}
+
+// Acquires the logging schema from the feature service, converts the memory buffer array of rows of logs and flushes
+// them to the offline storage.
+func (s *LoggingService) flushLogsToOfflineStorage(t time.Time) error {
+	log.Printf("Flushing buffer to offline storage with channel length: %d\n at time: "+t.String(), len(s.memoryBuffer.logs))
+	offlineStoreType, ok := getOfflineStoreType(s.fs.GetRepoConfig().OfflineStore)
+	if !ok {
+		return fmt.Errorf("could not get offline storage type for config: %s", s.fs.GetRepoConfig().OfflineStore)
+	}
+	if offlineStoreType == "file" {
+		entities, featureViews, odfvs, err := s.GetFcos()
+		if err != nil {
+			return err
+		}
+		schema, err := GetSchemaFromFeatureService(s.memoryBuffer.featureService, entities, featureViews, odfvs)
+		if err != nil {
+			return err
+		}
+		table, err := ConvertMemoryBufferToArrowTable(s.memoryBuffer, schema)
+		if err != nil {
+			return err
+		}
+		fileStore, err := NewFileOfflineStore(s.fs.GetRepoConfig().Project, s.fs.GetRepoConfig().OfflineStore)
+		if err != nil {
+			return err
+		}
+		fileStore.FlushToStorage(table)
+		s.memoryBuffer.logs = s.memoryBuffer.logs[:0]
+	} else {
+		// Currently don't support any other offline flushing.
+		return errors.New("currently only file type is supported for offline log storage")
+	}
+	return nil
+}
+
+// Takes memory buffer of logs in array row and converts them to columnar with generated fcoschema generated by GetFcoSchema
+// and writes them to arrow table.
+// Returns arrow table that contains all of the logs in columnar format.
+func ConvertMemoryBufferToArrowTable(memoryBuffer *MemoryBuffer, fcoSchema *Schema) (array.Table, error) {
+	arrowMemory := memory.NewGoAllocator()
+
+	columnNameToProtoValueArray := make(map[string][]*types.Value)
+	columnNameToStatus := make(map[string][]int32)
+	columnNameToTimestamp := make(map[string][]int64)
+	entityNameToEntityValues := make(map[string][]*types.Value)
+
+	for _, l := range memoryBuffer.logs {
+		// EntityTypes maps an entity name to the specific type and also which index in the entityValues array it is
+		// e.g if an Entity Key is {driver_id, customer_id}, then the driver_id entitytype would be dtype=int64, index=0.
+		// It's in the order of the entities as given by the schema.
+		for idx, entityName := range fcoSchema.Entities {
```

Comment:
on this point it's a `joinKey` not an `entityName`

Reply:
Fixed.